jena-arq/src/main/java/org/apache/jena/http/auth/AuthHeader.java

```diff
@@ -167,316 +193,407 @@ public String getBasicUserPassword() {
+            if ( idx < 0 )
+                return null;
+            if ( idx >= N )
+                return null;
+            StringBuilder stringBuilder = new StringBuilder();
+            // If reuse a string builder. -- stringBuilder.setLength(0);
+            char ch0 = nextChar();
+            if ( ch0 != '"' ) {
+                //System.err.println("Does not start with a dquote");
+                return null;
+            }
 
+            for (;;) {
+                char ch = nextChar();
+                if ( ch == EndOfStr )
+                    return null;
+                if ( ch == '"' )
+                    // Terminating double quote
+                    return stringBuilder.toString();
+                if ( ch == CH_RSLASH ) {
+                    ch = nextChar();
+                    if ( ch == EndOfStr )
+                        return null;
+                }
+                stringBuilder.append(ch);
+            }
+        }
 
-    //token68        = 1*( ALPHA / DIGIT /
-    //                     "-" / "." / "_" / "~" / "+" / "/" ) *"="
-    /** Single token68 */
-    private String token68() {
-        int j1 = whileTrue(string, idx, N, test_tok68);
-        int j2 = whileTrue(string, j1, N, test_base64_pad);
-        if ( j2 != N )
-            // Didn't reach the end of string.
-            return null;
-        return string.substring(idx,  j2);
-    }
+        // Base64 form defined in RFC4648, Section 4. with '+' and '/'
+        // An alternative is base64url uses '-' and '_'
+        private String basicBase64() {
+            int j1 = whileTrue(string, idx, N, test_base64);
+            int j2 = whileTrue(string, j1, N, test_base64_pad); // Strictly, one or two.
+            if ( j2 != N )
+                // Didn't reach the end of string.
+                return null;
+            return string.substring(idx,  j2);
+        }
 
-    // Bearer "base 64" URL encoding is given as
-    // b64token    = 1*( ALPHA / DIGIT /
-    //                  "-" / "." / "_" / "~" / "+" / "/" ) *"="
-    // which is token68 with trailing "=".
+        //token68        = 1*( ALPHA / DIGIT /
+        //                     "-" / "." / "_" / "~" / "+" / "/" ) *"="
+        /** RFC 7235 : Single token68 */
+        private String token68() {
+            int j1 = whileTrue(string, idx, N, test_tok68);
+            int j2 = whileTrue(string, j1, N, test_base64_pad);
+            if ( j2 != N )
+                // Didn't reach the end of string.
+                return null;
+            String s = string.substring(idx,  j2);
+            idx = j2;
+            return s;
+        }
 
-    private String b64token() { return token68(); }
+        // RFC6750 OAuth
+        // Bearer "base 64" URL encoding is given as
+        // b64token    = 1*( ALPHA / DIGIT /
+        //                  "-" / "." / "_" / "~" / "+" / "/" ) *"="
+        // which is token68.
+
+        private String b64token() { return token68(); }
```

Comment:
Yes - left for information. `b64token` is the rule name in RFC 6750 and has the same definition as `token68` in RFC 7235. 

But.

In the wild, JWT bearer tokens occur with padding of each of the three components making padding (uses of `=`) not just at the end as defined in the RFC's. This is reflected in `bearerToken()`, the code actually used in this PR.

This seems to be because some base64 decoders (JDK Base64, C# stdlib Base64, but not Apache Commons Codec - that decodes whether padded or not, as does JS) are sensitive to the base64 string being a multiple of 4 characters and require padding else generate an error. And some decoder need to know which variant of Base64 it is (original or URL safe - slightly different character sets) - again Apache Commons Codec just decodes both without fuss.

So padding in JWTs makes sense for robustness even if not quite RFC compliant.

And StackOverflow isn't helpful on this detail.

One RFC does mutter about "this is the defn but some auth schemes deviate from it".

I will not be surprised if we come back to this based on more real world situations.

Reply:
Interesting! I'm glad I press the button to expand resolved review comments. Let's hope we don't have any issues related to this, but if we do, at least we already have the code :crossed_fingers:  Thanks Andy!